src/SablierMerkleVCA.sol

```diff
@@ -117,40 +134,56 @@ contract SablierMerkleVCA is
     }
 
     /*//////////////////////////////////////////////////////////////////////////
-                          INTERNAL NON-CONSTANT FUNCTIONS
+                            INTERNAL CONSTANT FUNCTIONS
     //////////////////////////////////////////////////////////////////////////*/
 
-    /// @inheritdoc SablierMerkleBase
-    function _claim(uint256 index, address recipient, uint128 fullAmount) internal override {
-        uint40 blockTimestamp = uint40(block.timestamp);
-
+    /// @dev See the documentation for the user-facing functions that call this internal function.
+    function _calculateClaimAmount(uint128 fullAmount, uint40 claimTime) internal view returns (uint128) {
         // Load the vesting schedule into memory to avoid multiple SLOADs.
         MerkleVCA.Schedule memory vestingSchedule = _schedule;
 
-        // Check: schedule start time is in the past.
-        if (vestingSchedule.startTime >= blockTimestamp) {
-            revert Errors.SablierMerkleVCA_ClaimNotStarted(vestingSchedule.startTime);
-        }
-
-        // Calculate the claim and foregone amount.
-        uint128 claimAmount;
-        uint128 forgoneAmount;
-
-        if (vestingSchedule.endTime <= blockTimestamp) {
+        // Calculate the claim amount.
+        uint40 elapsedTime;
+        uint40 totalDuration;
+        if (claimTime >= vestingSchedule.endTime) {
             // If the vesting period has ended, the recipient can claim the full amount.
-            claimAmount = fullAmount;
+            return fullAmount;
         } else {
             // Otherwise, calculate the claimable amount based on the elapsed time.
-            uint40 elapsedTime = blockTimestamp - vestingSchedule.startTime;
-            uint40 totalDuration = vestingSchedule.endTime - vestingSchedule.startTime;
+            unchecked {
+                elapsedTime = claimTime - vestingSchedule.startTime;
+                totalDuration = vestingSchedule.endTime - vestingSchedule.startTime;
+            }
 
             // Safe to cast because the division results into a value less than `fullAmount`, which is already an
             // `uint128`.
-            claimAmount = uint128((uint256(fullAmount) * elapsedTime) / totalDuration);
+            return uint128((uint256(fullAmount) * elapsedTime) / totalDuration);
```

Comment:
Why not do this instead?

```suggestion
            return fullAmount * uint128(elapsedTime) / uint128(totalDuration);
```

Reply:
Because now both `elapsedTime` and `fullAmount` are `uint128`, the result would be of type `uint128`. However, during multiplication, the result can exceed `uint128` until the division is performed. Therefore, `fullAmount` was casted to `uint256` to accommodate multiplication result in case it exceeds `uint128`.  

So, its not dangerous cast.

Reply:
You're right!